collector/lib/CollectorConfig.cpp

```diff
@@ -447,6 +451,73 @@ void CollectorConfig::HandleConfig(const std::filesystem::path& filePath) {
   }
 }
 
+void WaitForFileToExist(const std::filesystem::path& filePath) {
+  while (!std::filesystem::exists(filePath)) {
+    sleep(5);
+  }
+}
+
+void CollectorConfig::WatchConfigFile(const std::filesystem::path& filePath) {
+  CLOG(INFO) << "In WatchConfigFile";
+  int fd = inotify_init();
+  if (fd < 0) {
+    CLOG(FATAL) << "inotify_init failed";
+    return;
+  }
+
+  WaitForFileToExist(filePath);
+  HandleConfig(filePath);
+  int wd = inotify_add_watch(fd, filePath.c_str(), IN_MODIFY | IN_MOVE_SELF | IN_DELETE_SELF);
+  if (wd < 0) {
+    CLOG(FATAL) << "Failed to add inotify watch for " << filePath;
+    close(fd);
+    return;
+  }
+
+  char buffer[1024];
+  while (true) {
+    CLOG(INFO) << "Waiting for file event";
+    int length = read(fd, buffer, sizeof(buffer));
+    if (length < 0) {
+      CLOG(FATAL) << "Unable to read event for " << filePath;
```

Comment:
I wouldn't kill collector completely just because of an inotify error, can we go into a state of trying to readd the watch instead?

Reply:
I have added a WaitForInotifyAddWatch function.